File: electron/sqlite/queries/contracts.js
/**
 * 合同管理 SQLite 查询模块
 *
 * 实现 income_contracts、income_records、expense_contracts、expense_records、
 * agreement_contracts 五张表的 CRUD 操作。
 * 采用工厂模式消除三种合同类型的重复代码。
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import log from 'electron-log';
import { tryGetSqlite, rowToCamel, toSqliteValue } from './helpers';
// ═══════════════════════════════════════════════════════════════════════════════
// 合同列映射（三种合同结构相同）
// ═══════════════════════════════════════════════════════════════════════════════
var CONTRACT_COLUMNS = {
    id: 'id',
    projectId: 'project_id',
    partnerId: 'partner_id',
    contractNo: 'contract_no',
    name: 'name',
    amount: 'amount',
    signedDate: 'signed_date',
    startDate: 'start_date',
    endDate: 'end_date',
    status: 'status',
    paymentMethod: 'payment_method',
    remarks: 'remarks',
    finalAmount: 'final_amount',
    settlementId: 'settlement_id',
    fileUrl: 'file_url',
    fileType: 'file_type',
    createdAt: 'created_at',
    updatedAt: 'updated_at',
};
// 其他协议多一个 agreement_type 字段
var AGREEMENT_EXTRA = {
    agreementType: 'agreement_type',
};
// ═══════════════════════════════════════════════════════════════════════════════
// 收入/支出记录列映射
// ═══════════════════════════════════════════════════════════════════════════════
var INCOME_RECORD_COLUMNS = {
    id: 'id',
    contractId: 'contract_id',
    amount: 'amount',
    recordDate: 'record_date',
    payer: 'payer',
    remarks: 'remarks',
    createdAt: 'created_at',
};
var EXPENSE_RECORD_COLUMNS = {
    id: 'id',
    contractId: 'contract_id',
    amount: 'amount',
    recordDate: 'record_date',
    payee: 'payee',
    remarks: 'remarks',
    createdAt: 'created_at',
};
// ═══════════════════════════════════════════════════════════════════════════════
// 辅助函数
// ═══════════════════════════════════════════════════════════════════════════════
function toInsertParams(columns, obj) {
    var insertCols = Object.values(columns).filter(function (c) { return c !== 'id'; });
    return insertCols.map(function (col) {
        var _a;
        var jsonKey = (_a = Object.entries(columns).find(function (_a) {
            var c = _a[1];
            return c === col;
        })) === null || _a === void 0 ? void 0 : _a[0];
        if (!jsonKey)
            return null;
        return toSqliteValue(obj[jsonKey]);
    });
}
function getInsertSQL(tableName, columns) {
    var insertCols = Object.values(columns).filter(function (c) { return c !== 'id'; });
    return "INSERT INTO \"".concat(tableName, "\" (").concat(insertCols.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(insertCols.map(function () { return '?'; }).join(', '), ")");
}
function toUpdateSet(columns, changes, excludeKeys) {
    if (excludeKeys === void 0) { excludeKeys = []; }
    var setClauses = [];
    var params = [];
    for (var _i = 0, _a = Object.entries(changes); _i < _a.length; _i++) {
        var _b = _a[_i], jsonKey = _b[0], value = _b[1];
        if (excludeKeys.includes(jsonKey))
            continue;
        var col = columns[jsonKey];
        if (!col)
            continue;
        setClauses.push("\"".concat(col, "\" = ?"));
        params.push(toSqliteValue(value));
    }
    return { sql: setClauses.join(', '), params: params };
}
var TABLE_MAP = {
    income: { contractTable: 'income_contracts', recordTable: 'income_records' },
    expense: { contractTable: 'expense_contracts', recordTable: 'expense_records' },
    agreement: { contractTable: 'agreement_contracts', recordTable: '' },
};
var COLUMNS_MAP = {
    income: CONTRACT_COLUMNS,
    expense: CONTRACT_COLUMNS,
    agreement: __assign(__assign({}, CONTRACT_COLUMNS), AGREEMENT_EXTRA),
};
var RECORD_COLUMNS_MAP = {
    income_records: INCOME_RECORD_COLUMNS,
    expense_records: EXPENSE_RECORD_COLUMNS,
};
/** 列出合同（含项目名、合作方名、收款/付款金额） */
export function listContracts(type, projectId) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var _b = TABLE_MAP[type], contractTable = _b.contractTable, recordTable_1 = _b.recordTable;
        var columns = COLUMNS_MAP[type];
        var sql = "SELECT c.*, p.name as project_name, pt.name as partner_name";
        if (recordTable_1) {
            var amountCol = type === 'income' ? 'received_amount' : 'paid_amount';
            sql += ", COALESCE(r.".concat(amountCol, ", 0) as computed_").concat(amountCol);
        }
        sql += " FROM \"".concat(contractTable, "\" c");
        sql += " LEFT JOIN projects p ON c.project_id = p.id";
        sql += " LEFT JOIN partners pt ON c.partner_id = pt.id";
        if (recordTable_1) {
            var amountCol = type === 'income' ? 'received_amount' : 'paid_amount';
            sql += " LEFT JOIN (SELECT contract_id, SUM(amount) as ".concat(amountCol, " FROM \"").concat(recordTable_1, "\" GROUP BY contract_id) r ON c.id = r.contract_id");
        }
        var params = [];
        if (projectId) {
            sql += ' WHERE c.project_id = ?';
            params.push(projectId);
        }
        sql += ' ORDER BY c.created_at DESC';
        var rows = (_a = sqlite.prepare(sql)).all.apply(_a, params);
        return rows.map(function (row) {
            var camel = rowToCamel(row);
            camel.projectName = row.project_name || '';
            camel.partnerName = row.partner_name || '';
            if (recordTable_1) {
                var amountKey = type === 'income' ? 'receivedAmount' : 'paidAmount';
                var computedKey = type === 'income' ? 'computed_received_amount' : 'computed_paid_amount';
                camel[amountKey] = row[computedKey] || 0;
                delete camel[computedKey.replace(/^computed_/, '')];
            }
            // 清理临时字段
            delete camel.project_name;
            delete camel.partner_name;
            return camel;
        });
    }
    catch (err) {
        log.error("[SQLite] contracts.list(".concat(type, ") error:"), err);
        return null;
    }
}
/** 创建合同 */
export function createContract(type, contract) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var contractTable = TABLE_MAP[type].contractTable;
        var columns = COLUMNS_MAP[type];
        var params = toInsertParams(columns, contract);
        var sql = getInsertSQL(contractTable, columns);
        (_a = sqlite.prepare(sql)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error("[SQLite] contracts.create(".concat(type, ") error:"), err);
        return false;
    }
}
/** 更新合同 */
export function updateContract(type, contract) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var contractTable = TABLE_MAP[type].contractTable;
        var columns = COLUMNS_MAP[type];
        var _b = toUpdateSet(columns, contract, ['id']), setSql = _b.sql, setParams = _b.params;
        if (!setSql)
            return true;
        setParams.push(new Date().toISOString());
        setParams.push(contract.id);
        var result = (_a = sqlite.prepare("UPDATE \"".concat(contractTable, "\" SET ").concat(setSql, ", \"updated_at\" = ? WHERE id = ?"))).run.apply(_a, setParams);
        return result.changes > 0;
    }
    catch (err) {
        log.error("[SQLite] contracts.update(".concat(type, ") error:"), err);
        return false;
    }
}
/** 删除合同（含级联删除记录） */
export function deleteContract(type, id) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var _a = TABLE_MAP[type], contractTable_1 = _a.contractTable, recordTable_2 = _a.recordTable;
        var doDelete = sqlite.transaction(function () {
            if (recordTable_2) {
                sqlite.prepare("DELETE FROM \"".concat(recordTable_2, "\" WHERE contract_id = ?")).run(id);
            }
            sqlite.prepare("DELETE FROM \"".concat(contractTable_1, "\" WHERE id = ?")).run(id);
        });
        doDelete();
        return true;
    }
    catch (err) {
        log.error("[SQLite] contracts.delete(".concat(type, ") error:"), err);
        return false;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// 收入/支出记录操作
// ═══════════════════════════════════════════════════════════════════════════════
/** 列出记录 */
export function listRecords(type, contractId) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var recordTable = TABLE_MAP[type].recordTable;
        var rows = sqlite.prepare("SELECT * FROM \"".concat(recordTable, "\" WHERE contract_id = ?")).all(contractId);
        return rows.map(rowToCamel);
    }
    catch (err) {
        log.error("[SQLite] records.list(".concat(type, ") error:"), err);
        return null;
    }
}
/** 创建记录 */
export function createRecord(type, record) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var recordTable = TABLE_MAP[type].recordTable;
        var columns = RECORD_COLUMNS_MAP[recordTable];
        var params = toInsertParams(columns, record);
        var sql = getInsertSQL(recordTable, columns);
        (_a = sqlite.prepare(sql)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error("[SQLite] records.create(".concat(type, ") error:"), err);
        return false;
    }
}
/** 删除记录 */
export function deleteRecord(type, id) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var recordTable = TABLE_MAP[type].recordTable;
        sqlite.prepare("DELETE FROM \"".concat(recordTable, "\" WHERE id = ?")).run(id);
        return true;
    }
    catch (err) {
        log.error("[SQLite] records.delete(".concat(type, ") error:"), err);
        return false;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// 合同统计
// ═══════════════════════════════════════════════════════════════════════════════
export function getContractStats() {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        // 收入合同总额
        var incomeRow = sqlite.prepare('SELECT COUNT(*) as count, COALESCE(SUM(COALESCE(final_amount, amount, 0)), 0) as total FROM income_contracts').get();
        // 支出合同总额
        var expenseRow = sqlite.prepare('SELECT COUNT(*) as count, COALESCE(SUM(COALESCE(final_amount, amount, 0)), 0) as total FROM expense_contracts').get();
        // 其他协议数量
        var agreementRow = sqlite.prepare('SELECT COUNT(*) as count FROM agreement_contracts').get();
        // 收款总额
        var incomeReceivedRow = sqlite.prepare("SELECT COALESCE(SUM(amount), 0) as total FROM payment_records WHERE type = 'invoice_out'").get();
        // 支出总额
        var expensePaidRow = sqlite.prepare("SELECT COALESCE(SUM(amount), 0) as total FROM payment_records WHERE type = 'invoice_in'").get();
        // 即将到期的合同（30天内）
        var now = new Date().toISOString().split('T')[0];
        var thirtyDaysLater = new Date(Date.now() + 30 * 24 * 60 * 60 * 1000).toISOString().split('T')[0];
        var expiringIncome = sqlite.prepare("SELECT id, 'income' as type, name, contract_no, COALESCE(final_amount, amount, 0) as amount, end_date, julianday(end_date) - julianday('now') as days_left FROM income_contracts WHERE end_date >= ? AND end_date <= ?").all(now, thirtyDaysLater);
        var expiringExpense = sqlite.prepare("SELECT id, 'expense' as type, name, contract_no, COALESCE(final_amount, amount, 0) as amount, end_date, julianday(end_date) - julianday('now') as days_left FROM expense_contracts WHERE end_date >= ? AND end_date <= ?").all(now, thirtyDaysLater);
        var expiringAgreement = sqlite.prepare("SELECT id, 'agreement' as type, name, contract_no, COALESCE(final_amount, amount, 0) as amount, end_date, julianday(end_date) - julianday('now') as days_left FROM agreement_contracts WHERE end_date >= ? AND end_date <= ?").all(now, thirtyDaysLater);
        var expiringSoon = __spreadArray(__spreadArray(__spreadArray([], expiringIncome.map(function (r) { return rowToCamel(r); }), true), expiringExpense.map(function (r) { return rowToCamel(r); }), true), expiringAgreement.map(function (r) { return rowToCamel(r); }), true).sort(function (a, b) { return a.daysLeft - b.daysLeft; });
        return {
            incomeCount: incomeRow.count,
            incomeTotal: incomeRow.total,
            incomeReceived: incomeReceivedRow.total,
            expenseCount: expenseRow.count,
            expenseTotal: expenseRow.total,
            expensePaid: expensePaidRow.total,
            agreementCount: agreementRow.count,
            netIncome: incomeRow.total - expenseRow.total,
            netReceived: incomeReceivedRow.total - expensePaidRow.total,
            expiringSoon: expiringSoon,
        };
    }
    catch (err) {
        log.error('[SQLite] contractStats.get error:', err);
        return null;
    }
}

File: src-tauri/src/commands/contracts.rs
/**
 * 合同管理命令
 *
 * 对应 Electron 版本的 contracts.ts
 * 包含收入合同、支出合同、其他协议
 */

use crate::db::AppState;
use crate::error::{AppError, AppResult};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{command, State};

// ============ 收入合同 ============

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IncomeContract {
    pub id: Option<i64>,
    pub project_id: Option<i64>,
    pub partner_id: Option<i64>,
    pub contract_no: Option<String>,
    pub name: Option<String>,
    pub amount: f64,
    pub signed_date: Option<String>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    pub status: Option<String>,
    pub payment_method: Option<String>,
    pub remarks: Option<String>,
    pub final_amount: Option<f64>,
    pub settlement_id: Option<i64>,
    pub file_url: Option<String>,
    pub file_type: Option<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
    pub project_name: Option<String>,
    pub partner_name: Option<String>,
}

/// 获取收入合同列表
#[command]
pub fn get_income_contracts(
    state: State<'_, AppState>,
    project_id: Option<i64>,
) -> AppResult<Vec<IncomeContract>> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let sql = if project_id.is_some() {
        "SELECT c.id, c.project_id, c.partner_id, c.contract_no, c.name, c.amount,
                c.signed_date, c.start_date, c.end_date, c.status, c.payment_method,
                c.remarks, c.final_amount, c.settlement_id, c.file_url, c.file_type,
                c.created_at, c.updated_at,
                p.name as project_name,
                pt.name as partner_name
         FROM income_contracts c
         LEFT JOIN projects p ON c.project_id = p.id
         LEFT JOIN partners pt ON c.partner_id = pt.id
         WHERE c.project_id = ?1
         ORDER BY c.created_at DESC"
    } else {
        "SELECT c.id, c.project_id, c.partner_id, c.contract_no, c.name, c.amount,
                c.signed_date, c.start_date, c.end_date, c.status, c.payment_method,
                c.remarks, c.final_amount, c.settlement_id, c.file_url, c.file_type,
                c.created_at, c.updated_at,
                p.name as project_name,
                pt.name as partner_name
         FROM income_contracts c
         LEFT JOIN projects p ON c.project_id = p.id
         LEFT JOIN partners pt ON c.partner_id = pt.id
         ORDER BY c.created_at DESC"
    };

    let mut stmt = db.prepare(sql).map_err(|e| AppError::Database(e.to_string()))?;

    let contracts = if let Some(pid) = project_id {
        stmt.query_map(params![pid], |row| {
            Ok(IncomeContract {
                id: Some(row.get(0)?),
                project_id: row.get(1)?,
                partner_id: row.get(2)?,
                contract_no: row.get(3)?,
                name: row.get(4)?,
                amount: row.get(5)?,
                signed_date: row.get(6)?,
                start_date: row.get(7)?,
                end_date: row.get(8)?,
                status: row.get(9)?,
                payment_method: row.get(10)?,
                remarks: row.get(11)?,
                final_amount: row.get(12)?,
                settlement_id: row.get(13)?,
                file_url: row.get(14)?,
                file_type: row.get(15)?,
                created_at: row.get(16)?,
                updated_at: row.get(17)?,
                project_name: row.get(18)?,
                partner_name: row.get(19)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    } else {
        stmt.query_map([], |row| {
            Ok(IncomeContract {
                id: Some(row.get(0)?),
                project_id: row.get(1)?,
                partner_id: row.get(2)?,
                contract_no: row.get(3)?,
                name: row.get(4)?,
                amount: row.get(5)?,
                signed_date: row.get(6)?,
                start_date: row.get(7)?,
                end_date: row.get(8)?,
                status: row.get(9)?,
                payment_method: row.get(10)?,
                remarks: row.get(11)?,
                final_amount: row.get(12)?,
                settlement_id: row.get(13)?,
                file_url: row.get(14)?,
                file_type: row.get(15)?,
                created_at: row.get(16)?,
                updated_at: row.get(17)?,
                project_name: row.get(18)?,
                partner_name: row.get(19)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    };

    Ok(contracts)
}

/// 创建收入合同
#[command]
pub fn create_income_contract(
    state: State<'_, AppState>,
    contract: IncomeContract,
) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    db.execute(
        "INSERT INTO income_contracts (project_id, partner_id, contract_no, name, amount,
         signed_date, start_date, end_date, status, payment_method, remarks)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)",
        params![
            contract.project_id,
            contract.partner_id,
            contract.contract_no,
            contract.name,
            contract.amount,
            contract.signed_date,
            contract.start_date,
            contract.end_date,
            contract.status,
            contract.payment_method,
            contract.remarks,
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    let id = db.last_insert_rowid();
    Ok(id)
}

/// 更新收入合同
#[command]
pub fn update_income_contract(
    state: State<'_, AppState>,
    contract: IncomeContract,
) -> AppResult<()> {
    let id = contract.id.ok_or_else(|| AppError::Validation("更新收入合同需要 id".into()))?;
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute(
            "UPDATE income_contracts SET project_id = ?1, partner_id = ?2, contract_no = ?3,
             name = ?4, amount = ?5, signed_date = ?6, start_date = ?7, end_date = ?8,
             status = ?9, payment_method = ?10, remarks = ?11, final_amount = ?12,
             settlement_id = ?13, updated_at = datetime('now')
             WHERE id = ?14",
            params![
                contract.project_id,
                contract.partner_id,
                contract.contract_no,
                contract.name,
                contract.amount,
                contract.signed_date,
                contract.start_date,
                contract.end_date,
                contract.status,
                contract.payment_method,
                contract.remarks,
                contract.final_amount,
                contract.settlement_id,
                id,
            ],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!(
            "收入合同 {} 不存在",
            id
        )));
    }

    Ok(())
}

/// 删除收入合同
#[command]
pub fn delete_income_contract(state: State<'_, AppState>, id: i64) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute("DELETE FROM income_contracts WHERE id = ?1", params![id])
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("收入合同 {} 不存在", id)));
    }

    Ok(())
}

// ============ 支出合同（结构类似，省略详细实现） ============

/// 支出合同类型（复用收入合同结构）
pub type ExpenseContract = IncomeContract;

/// 获取支出合同列表
#[command]
pub fn get_expense_contracts(
    state: State<'_, AppState>,
    project_id: Option<i64>,
) -> AppResult<Vec<ExpenseContract>> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let sql = if project_id.is_some() {
        "SELECT c.id, c.project_id, c.partner_id, c.contract_no, c.name, c.amount,
                c.signed_date, c.start_date, c.end_date, c.status, c.payment_method,
                c.remarks, c.final_amount, c.settlement_id, c.file_url, c.file_type,
                c.created_at, c.updated_at,
                p.name as project_name,
                pt.name as partner_name
         FROM expense_contracts c
         LEFT JOIN projects p ON c.project_id = p.id
         LEFT JOIN partners pt ON c.partner_id = pt.id
         WHERE c.project_id = ?1
         ORDER BY c.created_at DESC"
    } else {
        "SELECT c.id, c.project_id, c.partner_id, c.contract_no, c.name, c.amount,
                c.signed_date, c.start_date, c.end_date, c.status, c.payment_method,
                c.remarks, c.final_amount, c.settlement_id, c.file_url, c.file_type,
                c.created_at, c.updated_at,
                p.name as project_name,
                pt.name as partner_name
         FROM expense_contracts c
         LEFT JOIN projects p ON c.project_id = p.id
         LEFT JOIN partners pt ON c.partner_id = pt.id
         ORDER BY c.created_at DESC"
    };

    let mut stmt = db.prepare(sql).map_err(|e| AppError::Database(e.to_string()))?;

    let contracts = if let Some(pid) = project_id {
        stmt.query_map(params![pid], |row| {
            Ok(IncomeContract {
                id: Some(row.get(0)?),
                project_id: row.get(1)?,
                partner_id: row.get(2)?,
                contract_no: row.get(3)?,
                name: row.get(4)?,
                amount: row.get(5)?,
                signed_date: row.get(6)?,
                start_date: row.get(7)?,
                end_date: row.get(8)?,
                status: row.get(9)?,
                payment_method: row.get(10)?,
                remarks: row.get(11)?,
                final_amount: row.get(12)?,
                settlement_id: row.get(13)?,
                file_url: row.get(14)?,
                file_type: row.get(15)?,
                created_at: row.get(16)?,
                updated_at: row.get(17)?,
                project_name: row.get(18)?,
                partner_name: row.get(19)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    } else {
        stmt.query_map([], |row| {
            Ok(IncomeContract {
                id: Some(row.get(0)?),
                project_id: row.get(1)?,
                partner_id: row.get(2)?,
                contract_no: row.get(3)?,
                name: row.get(4)?,
                amount: row.get(5)?,
                signed_date: row.get(6)?,
                start_date: row.get(7)?,
                end_date: row.get(8)?,
                status: row.get(9)?,
                payment_method: row.get(10)?,
                remarks: row.get(11)?,
                final_amount: row.get(12)?,
                settlement_id: row.get(13)?,
                file_url: row.get(14)?,
                file_type: row.get(15)?,
                created_at: row.get(16)?,
                updated_at: row.get(17)?,
                project_name: row.get(18)?,
                partner_name: row.get(19)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?
    };

    Ok(contracts)
}

/// 创建支出合同
#[command]
pub fn create_expense_contract(
    state: State<'_, AppState>,
    contract: ExpenseContract,
) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    db.execute(
        "INSERT INTO expense_contracts (project_id, partner_id, contract_no, name, amount,
         signed_date, start_date, end_date, status, payment_method, remarks)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)",
        params![
            contract.project_id,
            contract.partner_id,
            contract.contract_no,
            contract.name,
            contract.amount,
            contract.signed_date,
            contract.start_date,
            contract.end_date,
            contract.status,
            contract.payment_method,
            contract.remarks,
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    let id = db.last_insert_rowid();
    Ok(id)
}

/// 更新支出合同
#[command]
pub fn update_expense_contract(
    state: State<'_, AppState>,
    contract: ExpenseContract,
) -> AppResult<()> {
    let id = contract.id.ok_or_else(|| AppError::Validation("更新支出合同需要 id".into()))?;
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute(
            "UPDATE expense_contracts SET project_id = ?1, partner_id = ?2, contract_no = ?3,
             name = ?4, amount = ?5, signed_date = ?6, start_date = ?7, end_date = ?8,
             status = ?9, payment_method = ?10, remarks = ?11, final_amount = ?12,
             settlement_id = ?13, updated_at = datetime('now')
             WHERE id = ?14",
            params![
                contract.project_id,
                contract.partner_id,
                contract.contract_no,
                contract.name,
                contract.amount,
                contract.signed_date,
                contract.start_date,
                contract.end_date,
                contract.status,
                contract.payment_method,
                contract.remarks,
                contract.final_amount,
                contract.settlement_id,
                id,
            ],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!(
            "支出合同 {} 不存在",
            id
        )));
    }

    Ok(())
}

/// 删除支出合同
#[command]
pub fn delete_expense_contract(state: State<'_, AppState>, id: i64) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute("DELETE FROM expense_contracts WHERE id = ?1", params![id])
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("支出合同 {} 不存在", id)));
    }

    Ok(())
}

// ============ 合同统计 ============

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ContractStats {
    pub income_count: i64,
    pub income_total: f64,
    pub income_received: f64,
    pub expense_count: i64,
    pub expense_total: f64,
    pub expense_paid: f64,
    pub agreement_count: i64,
    pub net_income: f64,
    pub net_received: f64,
}

/// 获取合同统计
#[command]
pub fn get_contract_stats(state: State<'_, AppState>) -> AppResult<ContractStats> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let income_count: i64 = db
        .query_row("SELECT COUNT(*) FROM income_contracts", [], |row| row.get(0))
        .map_err(|e| AppError::Database(e.to_string()))?;

    let income_total: f64 = db
        .query_row(
            "SELECT COALESCE(SUM(amount), 0) FROM income_contracts",
            [],
            |row| row.get(0),
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    let expense_count: i64 = db
        .query_row(
            "SELECT COUNT(*) FROM expense_contracts",
            [],
            |row| row.get(0),
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    let expense_total: f64 = db
        .query_row(
            "SELECT COALESCE(SUM(amount), 0) FROM expense_contracts",
            [],
            |row| row.get(0),
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    let agreement_count: i64 = db
        .query_row(
            "SELECT COUNT(*) FROM agreement_contracts",
            [],
            |row| row.get(0),
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    // 计算已收款（从 payment_records）
    let income_received: f64 = db
        .query_row(
            "SELECT COALESCE(SUM(amount), 0) FROM payment_records WHERE type = 'invoice_out'",
            [],
            |row| row.get(0),
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    let expense_paid: f64 = db
        .query_row(
            "SELECT COALESCE(SUM(amount), 0) FROM payment_records WHERE type = 'invoice_in'",
            [],
            |row| row.get(0),
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(ContractStats {
        income_count,
        income_total,
        income_received,
        expense_count,
        expense_total,
        expense_paid,
        agreement_count,
        net_income: income_total - expense_total,
        net_received: income_received - expense_paid,
    })
}
